HeterogeneousCore/SonicTriton/src/TritonClient.cc

```diff
@@ -0,0 +1,318 @@
+}
+
+template <typename Client>
+bool TritonClient<Client>::wrap(const nic::Error& err, const std::string& msg) const {
+  if (!err.IsOk()) {
+    //emit warning
+    edm::LogWarning("TritonServerWarning") << msg << ": " << err;
+  }
+  return err.IsOk();
+}
+
+template <typename Client>
+bool TritonClient<Client>::setup() {
+  bool status = true;
+
+  status = wrap(nic::InferGrpcContext::Create(&context_, url_, modelName_, modelVersion_, false),
+                "setup(): unable to create inference context");
+  if (!status)
+    return status;
+
+  //only used for monitoring
+  bool has_server = false;
+  if (verbose_) {
+    bool has_server = wrap(nic::ServerStatusGrpcContext::Create(&serverCtx_, url_, false),
+                           "setup(): unable to create server context");
+  }
+  if (!has_server)
+    serverCtx_ = nullptr;
+
+  std::unique_ptr<nic::InferContext::Options> options;
+  status = wrap(nic::InferContext::Options::Create(&options), "setup(): unable to create inference context options");
+  if (!status)
+    return status;
+
+  options->SetBatchSize(batchSize_);
+  for (const auto& output : context_->Outputs()) {
+    status = wrap(options->AddRawResult(output), "setup(): unable to add raw result");
+    if (!status)
+      return status;
+  }
+  status = wrap(context_->SetRunOptions(*options), "setup(): unable to set run options");
+  if (!status)
+    return status;
+
+  const auto& nicInputs = context_->Inputs();
+  nicInput_ = nicInputs[0];
+  nicInput_->Reset();
+
+  auto t1 = std::chrono::high_resolution_clock::now();
+  std::vector<int64_t> input_shape;
+  for (unsigned i0 = 0; i0 < batchSize_; i0++) {
+    float* arr = &(this->input_.data()[i0 * nInput_]);
+    status = wrap(nicInput_->SetRaw(reinterpret_cast<const uint8_t*>(arr), nInput_ * sizeof(float)),
+                  "setup(): unable to set data for batch entry " + std::to_string(i0));
+    if (!status)
+      return status;
+  }
+  auto t2 = std::chrono::high_resolution_clock::now();
+  if (!this->debugName_.empty())
+    edm::LogInfo(this->fullDebugName_) << "Input array time: "
+                                       << std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
+
+  //should be true
+  return status;
+}
+
+template <typename Client>
+bool TritonClient<Client>::getResults(const std::unique_ptr<nic::InferContext::Result>& result) {
+  bool status = true;
+
+  auto t1 = std::chrono::high_resolution_clock::now();
+  this->output_.resize(nOutput_ * batchSize_, 0.f);
+  for (unsigned i0 = 0; i0 < batchSize_; i0++) {
+    const uint8_t* r0;
+    size_t content_byte_size;
+    status = wrap(result->GetRaw(i0, &r0, &content_byte_size),
+                  "getResults(): unable to get raw for entry " + std::to_string(i0));
+    if (!status)
+      return status;
+    const float* lVal = reinterpret_cast<const float*>(r0);
+    std::memcpy(&this->output_[i0 * nOutput_], lVal, nOutput_ * sizeof(float));

```

Comment:
Given that the copy goes via `memcpy()`, the `reinterpret_cast` above is unnecessary, no? (`std::copy()` would be different)